force-app/main/default/classes/SOQL.cls

```diff
@@ -344,18 +409,43 @@ public with sharing class SOQL {
      * Example: setPrimaryContact() should set the 'Primary_Contact__c' field for every Opportunity in the org
      */
     public static void setPrimaryContact() {
-        //Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
+        // Step 1: Use the getOpportunityDetailsForClosedWon() method to get the Opportunities
+        List<Opportunity> closedWonOpps = getOpportunityDetailsForClosedWon();
 
-        // Create a set of Account Ids from the Opportunities
-
-        // Query the Contacts
-
-        // Loop through each Contact and add the Account Id and Contact to the map
-        
-        // Loop through each Opportunity
-
-                // Update the Opportunity primary contact from accountToContactMap
-
-        // Update the Opportunities
+        // Step 2: Create a map to store the primary contact for each Account
+        Map<Id, Id> accountToFirstContactMap = new Map<Id, Id>();
+    
+        // Step 3: Loop through closedWonOpps to collect Account IDs
+        for (Opportunity opp : closedWonOpps) {
+            if (opp.AccountId != null && !accountToFirstContactMap.containsKey(opp.AccountId)) {
+                accountToFirstContactMap.put(opp.AccountId, null); // Initialize with null
+            }
+        }
+    
+            // Step 4: Query Contacts for each Account and populate accountToFirstContactMap with the logically first ContactId
+        for (Contact contact : [
```

Comment:
You could eliminate this loop by querying for Accounts with related Contacts and assigning directly to a map